### PR TITLE
Iox 1815 add determinism label to selected types

### LINKED ISSUE #1815: Label deterministic functions
## Brief feature description

Label all functions that are intended to be used at runtime safely as `\deterministic` (or maybe @deterministic).
This is intended to be additional user information and used by code analysis tools.

Deterministic is a little bit of a misnomer unfortunately as it is not meaning what one might expect here (deterministic runtime/results)... We will maybe later switch this label to `@runtime-safe`, the labeling right now is done for tooling purposes.

## Detailed information

The labeling can be carried out in batches (i.e. logical units of files), almost all functions not relying on POSIX resources will be marked.

### PR DESCRIPTION
**Pull request by MatthiasKillat**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## References

- Closes #1815

### CODE REVIEW

File: doc/design/determinism.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# Determinism in iceoryx
+
+Deterministic exectution is important for many application domains, especially real-time systems.
```
Comment (MatthiasKillat) [inline doc/design/determinism.md:3]:
```suggestion
Deterministic execution is important for many application domains, especially real-time systems.
```

File: doc/design/determinism.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+output.
+
+The state of the environment refers to any global or member variables the function depends on.
+In particular it may depend on the current time or a pseudo random number generator state.
+
+This is the (informal) definition used in computer science.
+
+## Deterministic runtime
+
+The runtime of a function is bounded and the function cannot block indefinitely under normal
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/determinism.md:21]: What are normal conditions?

File: doc/design/determinism.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+conditions. It must be guaranteed that the function execution is scheduled by the underlying
+OS scheduler.
+
+However, while such time-bounds are guaranteed to exist they are usually not known
+and depend on hardware and generally the system state (i.e. other processes, context switches etc.).
+
+This is definition is mostly used in real-time systems theory and in the
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/determinism.md:28]:
```suggestion
This definition is mostly used in real-time systems theory and in the
```

File: doc/design/determinism.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+following referred to as time-determinism.
+
+## Influences on time-determinism
+
+1. Thread timing (including benign race conditions)
+1. Context switches
+1. Branch prediction and speculative exectution
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/determinism.md:35]:
```suggestion
1. Branch prediction and speculative execution
```

File: doc/design/determinism.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+Except for functions whose purpose is blocking, `\deterministc` guarantees time-determinism in the
+above sense. This makes `\deterministic` functions suitable to be used in real-time applications
+where a (somewhat) predictable runtime is required.
+
+We only consider public API for annotation, private functions are implementation details and do 
+not require annotation (it can be considered optionally where it is useful).
+
+## Conditional guarantees for templates
+
+iceoryx hoofs provides many template classes and functions functions that depend on user-defined
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/determinism.md:69]:
```suggestion
iceoryx hoofs provides many template classes and functions that depend on user-defined
```

File: doc/design/determinism.md (hunk 2 above)
Comment (MatthiasKillat) [reply doc/design/determinism.md:21]: Good question, I will remove it or specify it. I think it might be helpful to allow infinite blocking in abnormal conditions, which for me means a system failure.

Say we trigger a fatal error, from then on all bets are of and the function will possibly not return. This is also a design philosophy in our error handling (currently and will be extended). We call terminate, but the function does not return. Would this be called blocking? Also, it is not guaranteed that termination happens "immediately" etc.

There is also the possibility to send the system into an infinite loop during FATAL errors (some design approaches do this, as kind of a safe state). This would also be possible in the future.

TLDR: I would like to state that during FATAL errors no runtime bound can be assumed.

**Review by elBoberido — COMMENTED** [review]
Looks good overall. Can be merged once the minor findings are fixed.

File: doc/design/determinism.md (hunk 2 above)
Comment (MatthiasKillat) [reply doc/design/determinism.md:21]: Clarified?

**Comment by MatthiasKillat** [comment]
@elBoberido @FerdinandSpitzschnueffler Reduced the scope to the two experimental modules as discussed. All functions are "deterministic" there, as in many other modules (except for POSIX and arguably some concurrency stuff).

Still, we need more experience with the test validation and I documented the reason for scope reduction to make it clear to users that this is not fully rolled out for good reason.

I can squash all the commits if desired (a better solution, as nothing much is happening). Let me know if this is OK before or after review.

File: doc/design/determinism.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+Currently this labeling is experimental due to interaction to gain experience with automatic 
+validation methods that verify whether a function is indeed `@determinisitic`. 
+Since the validation interferes with tests, the labeling is currently limited to a few headers only. 
+It will be expanded in the future if it proves useful and feasible.
+
+1. Reevaluate dynamic memory allocation guarantee for POSIX calls.
+1. `@deterministic` functions could be more precisely annotated as time-bounded or real-time-safe.
+1. Provide more detailed guarantees such as non-blocking or non-allocating.
+1. Extend the guarantee for polymorphic classes (similar restrictions as with callbacks).
+1. Use `@deterministic` annotation instead.
```
Comment (FerdinandSpitzschnueffler) [inline doc/design/determinism.md:114]: Can be removed now.

File: iceoryx_hoofs/container/include/iox/vector.hpp [hunk 7]
```diff
@@ -43,96 +43,115 @@ class vector final
...
     /// @param[in] index of the element to return
     /// @return const reference to the element stored at index
     /// @attention Out of bounds access leads to a program termination!
+    /// @deterministic
     const T& operator[](const uint64_t index) const noexcept;
 
     /// @brief returns a reference to the first element; terminates if the vector is empty
     /// @return reference to the first element
     /// @attention Accessing an empty vector leads to a program termination!
+    /// @deterministic
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/container/include/iox/vector.hpp:154]: What about the const front method?

Comment (MatthiasKillat) [reply iceoryx_hoofs/container/include/iox/vector.hpp:154]: Oversight. One really needs tooling for this e.g. based on LLVM.

**Review by FerdinandSpitzschnueffler — DISMISSED** [review]

File: doc/design/determinism.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+the determinism guarantee themselves.
+
+Note that for example the `size` function of `vector` does not depend on the template type `T`, i.e.
+the set of relevant functions above is empty. In this case the function is unconditionally
+`@deterministic`.
+
+## Conditional guarantees for callbacks
+
+For callbacks specified by the user and provided as e.g. lambda expressions, function pointers or
+function wrappers such as `cxx::function` or `std::function` the `@deterministic` guarantee is also
```
Comment (elBoberido) [inline doc/design/determinism.md:91]:
```suggestion
function wrappers such as `iox::function` or `std::function` the `@deterministic` guarantee is also
```

**Review by elBoberido — APPROVED** [review]
This minor naming issue could also be solved in another PR. We will need to grep for this namespaces anyway.

**Review by FerdinandSpitzschnueffler — APPROVED** [review]
